[DOC] Add dissect processor documentation

vagimeli opened the pull request:
### Description
Address content gap for ingest processors documentation. Adds new content for dissect processor.

### Issues Resolved
Fixes #4647 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

vagimeli commented:
@heemin32 Please review the new dissect processor when you have availability. Thank you, Melissa

heemin32 commented:
> @heemin32 Please review the new dissect processor when you have availability. Thank you, Melissa

Will try to review it but I am fully occupied with other works. Could you also check with others to see if they can review it?

vagimeli commented:
> > @heemin32 Please review the new dissect processor when you have availability. Thank you, Melissa
> 
> Will try to review it but I am fully occupied with other works. Could you also check with others to see if they can review it?

Yes, will do. Thanks Heemin :)

vagimeli commented:
> @vamshin This PR is ready for technical review by your team. Thanks for helping move it forward :)
> 
> cc: @hdhalter for visibility

Hi @vamshin, Do we have a dev who can review this content for technical accuracy? Thanks, Melissa

vamshin commented:
> > @vamshin This PR is ready for technical review by your team. Thanks for helping move it forward :)
> > cc: @hdhalter for visibility
> 
> Hi @vamshin, Do we have a dev who can review this content for technical accuracy? Thanks, Melissa

@vagimeli not sure the right dev for this.  Adding Search pipelines dev for insights @macohen @msfroh

vagimeli commented:
Hi @macohen & @msfroh, Will you review, or direct me to the appropriate reviewer, for the ingest processor documentation. We're working through content gaps and need dev reviewers to vet technical accuracy. Thank you.

msfroh commented:
@vagimeli I can take a look. Thanks!

vagimeli commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,281 @@
+
+`value1, value2`
+
+### Append with order modifier (`+` and `/n`)
```
@msfroh Thank you for this comment. Will you update the text to read accurately. I'm not technical in this area, so your expertise would greatly help. Thank you.

vagimeli commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,281 @@
+|-----------|-----------|-----------|
+`->` | Skip right padding | (far) right | `%{keyname->}` | Tells `dissect` to skip over any repeated characters to the right. For example, `%{timestamp->}` could be used to tell `dissect` to skip over any padding characters, such as two spaces or any varying character padding, that follow `timestamp`. |
+`+` | Append | left | `%{keyname} %{+keyname}` | Appends two or more fields together. | 
+`+` with `/n` | Append with order | left and right | `%{+keyname}/2 %{+keyname/1}` | Appends two or more fields together in the order specified. |
```
@msfroh: Is `/n` accurate here?

vagimeli commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,281 @@
+
+The append with order modifier combines the values of two or more keys into a single output value, adhering to the a specific order defines by a newline character `/n`. You have the flexibility to customize the separator that separates the appended values. the append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items,  ...
+
+#### Example
```
@msfroh Please confirm I've properly interpreted your comment and example.

vagimeli commented:
@kolchfa-aws This can be added to the doc review queue at your availability.

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+# Dissect
+
+The `dissect` processor extracts values from an event and maps them to individual fields based on user-defined dissect patterns. The processor is well-suited for field extractions from log messages with a known structure.
```
```suggestion
The `dissect` processor extracts values from a document text field and maps them to individual fields based on dissect patterns. The processor is well-suited for field extractions from log messages with a known structure. Unlinke the `grok` processor, `dissect` does not use regular expressions and has a simpler syntax.
```

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+#### Example of usage
+
+The following is an example of how to use the append modifier. In this example, the pattern extracts the values of `key1` and `key2` fields and appends them together, with a space as the separator:
```
The field name to which you are appending has to be the same for all. Remove the `key1` and `key2` example and add the following example:

```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{+address}, %{+address} %{+address}",
        "append_separator": "|"
      }
    }
  ]
}
```
Test pipeline:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "New York, NY 10017"
      }
    }
  ]
}
```
In the response, the substrings are appended into the `address` field:

```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "address": "New York|NY|10017",
          "message": "New York, NY 10017"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:30:54.516284637Z"
        }
      }
    }
  ]
}
```

vagimeli replied:
Thank you for the example.

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+`value1 value2`
+
+You can also specify a custom separator using the `append_separator` parameter. For example, the following pattern uses a comma as the separator: 
```
This syntax is incorrect. Remove this part.

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+### Append with order modifier (`+` and `/`)
+
+The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/`. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values fr ...
```
This section does not show an order modifier. Here's an example that does:

Specify reverse order for the previous example
```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{+address/3}, %{+address/2} %{+address/1}",
        "append_separator": "|"
      }
    }
  ]
}
```
Test the pipeline with the same request:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "New York, NY 10017"
      }
    }
  ]
}
```
The substrings are appended into the `address` fields in reverse order:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "address": "10017|NY|New York",
          "message": "New York, NY 10017"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:38:24.305974178Z"
        }
      }
    }
  ]
}
```

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+### Named skip key
+
+The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output, focusing on specific information ...
```
```suggestion
The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output. 
```

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+#### Example of usage
+
+The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final result:
```
How about a full example:

```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{firstName} %{lastName} %{?ignore}"
      }
    }
  ]
}
```
Test:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "John Doe M.D."
      }
    }
  ]
}
```
Response:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "firstName": "John",
          "lastName": "Doe",
          "message": "John Doe M.D."
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:41:58.161475555Z"
        }
      }
    }
  ]
}
```

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,289 @@
+
+#### Example of usage
+
+The following pattern uses a reference key to extract data into a structured format. In this example, `%{value}` represents the parsed value and `%{reference_key}` acts as the placeholder for the actual key:
```
Here's an example:
extract `client_ip` and two key/value pairs for the next values:
```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{client_ip} %{*a}:%{&a} %{*b}:%{&b}"
      }
    }
  ]
}
```
Test the pipeline:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "192.168.1.10 response_code:200 response_size:3456"
      }
    }
  ]
}
```
The two key/value pairs were extracted into fields:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "client_ip": "192.168.1.10",
          "response_code": "200",
          "message": "192.168.1.10 response_code:200 response_size:3456",
          "response_size": "3456"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:48:51.475535635Z"
        }
      }
    }
  ]
}
```

vagimeli replied:
Thank you for the examples.

kolchfa-aws reviewed (changes requested):
Nice work. Please update the examples according to my comments. Thanks!

vagimeli commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,289 @@
+New York City, NY 10017
+```
+
+Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries, even though they have slightly different formats.
````
thank you for the example :)

vagimeli commented:
@kolchfa-aws Thank you for the thorough review and the many examples. I've accepted all of your edits and addressed your comments. Please re-review at your availability. Thank you :)

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+### Append with order modifier (`+` and `/`)
````
```suggestion
### Append with order modifier (`+` and `/n`)
```

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+### Append with order modifier (`+` and `/`)
+
+The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/` modifier. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating  ...
```
```suggestion
The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/`. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values from various sources.  
```

kolchfa-aws commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+#### Example of usage
+
+The following example pipeline uses the append with order modifier to reverse the pattern order defined in the preceding pipeline. The `append_separator` parameter must be defined in the processor configuration, outside of the `pattern`. It is only relevant with the `+` modifier.
```
```suggestion
The following example pipeline uses the append with order modifier to reverse the pattern order defined in the preceding pipeline. This pipeline specifies a separator to insert between the appended fields. If you don't specify a separator, all values will be appended together without a separator.
```

kolchfa-aws reviewed (changes requested):
Just a couple of more changed. Thank you!

vagimeli commented:
> Just a couple of more changed. Thank you!

@kolchfa-aws Accepted your edits. Thanks so much for your time and help with these processor docs :)

kolchfa-aws reviewed (approved):
Thank you, @vagimeli! Nice job. LGTM

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+The right padding modifier is used to allow for the repetition of characters following a `%{keyname->}`. The right padding modifier can be applied to any key along with any other modifiers. It should always be the rightmost modifier, for example, `%{+keyname/1->}` or `%{}`.
+
+#### Example of usage
```
```suggestion
#### Example usage
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+`%{city->}, %{state} %{zip}`
+
+In this pattern, the right padding modifier `->` is applied to the `%{city}` key. Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries, even though they have slightly different formats: 
```
```suggestion
In this pattern, the right padding modifier `->` is applied to the `%{city}` key. Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries even though they have slightly different formats: 
```

natebower replied:
Should "city" be in code font?

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+You can test the pipeline using the following example:
````
```suggestion
You can test the pipeline by using the following example:
```

natebower replied:
Can we add the noun after "example"?

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+Your response should be similar to the following:
````
```suggestion
Your response should appear similar to the following:
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+### Append modifier (`+`)
+
+The append modifier combines the values of two or more values into a single output value. The values are appended from left to right. You can also specify an optional separator to be inserted between the values. 
```
Global: Should the names of the modifiers be in code font or otherwise formatted to indicate that they are names?

natebower replied:
In that case, it should be "An append modifier".

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+The append modifier combines the values of two or more values into a single output value. The values are appended from left to right. You can also specify an optional separator to be inserted between the values. 
+
+#### Example of usage
```
```suggestion
#### Example usage
```

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+You can test the pipeline using the following example:
````
```suggestion
You can test the pipeline by using the following example:
```

natebower replied:
Can we add the noun after "example"?

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/`. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values fr ...
+
+#### Example of usage
```
```suggestion
#### Example usage
```

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+### Append with order modifier (`+` and `/n`)
````
Re: my previous comments, this modifier name is particularly awkward without some way of indicating that it's a name.

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+You can test the pipeline using the following example:
````
Can we add the noun after "example"?

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+The substrings are appended into the `address` field in reverse order, as shown in the following response:
````
```suggestion
The substrings are appended to the `address` field in reverse order, as shown in the following response:
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+### Named skip key
+
+The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output.
```
```suggestion
The named skip key modifier excludes specific matches from the final output by using an empty key (`{}` or `?`) modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output.
```

natebower replied:
This modifier name is also very awkward without some formatting to indicate that it's a name.

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output.
+
+#### Example of usage
```
```suggestion
#### Example usage
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+#### Example of usage
+
+The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final result:
```
```suggestion
The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final output:
```

natebower replied:
Should "modifier" follow "named skip key"?

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+You can test the pipeline using the following example:
````
```suggestion
You can test the pipeline by using the following example:
```

natebower replied:
Can we add the noun after "example"?

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+Your response should be similar to the following:
````
```suggestion
Your response should appear similar to the following:
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+### Reference keys (`*` and `&`)
+
+Reference keys use parsed values as key/value pairings for structured content. This can use useful when handling systems that partially log data in key/value pairs. By using reference keys, you can preserve the key/value relationship and maintain the integrity of the extracted information. 
```
```suggestion
Reference keys use parsed values as key-value pairings for structured content. This can use useful when handling systems that partially log data in key-value pairs. By using reference keys, you can preserve the key-value relationship and maintain the integrity of the extracted information. 
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+Reference keys use parsed values as key/value pairings for structured content. This can use useful when handling systems that partially log data in key/value pairs. By using reference keys, you can preserve the key/value relationship and maintain the integrity of the extracted information. 
+
+#### Example of usage
```
```suggestion
#### Example usage
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+#### Example of usage
+
+The following pattern uses a reference key to extract data into a structured format. In this example, `client_ip` and two key/value pairs are extracted for the next values:
```
```suggestion
The following pattern uses a reference key to extract data into a structured format. In this example, `client_ip` and two key-value pairs are extracted for the next values:
```

natebower commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+#### Example of usage
+
+The following pattern uses a reference key to extract data into a structured format. In this example, `client_ip` and two key/value pairs are extracted for the next values:
```
"extract...into" doesn't work. Please revise. Do we mean something like "to extract data for use in a structured format"?

natebower replied:
Follow-up: @kolchfa-aws has indicated that "extract into" is common usage in a CS context. I would still prefer different/clearer wording but will defer to you.

vagimeli replied:
Revised to read: "to find data and organize data it into a more structured format."

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+You can test the pipeline using the following example:
````
```suggestion
You can test the pipeline by using the following example:
```

natebower replied:
Can we add the noun after "example"?

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+The two key/value pairs were extracted into fields, as shown in the following response:
````
```suggestion
The two key-value pairs were extracted into fields, as shown in the following response:
```

natebower commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+```
+{% include copy-curl.html %}
+
+The two key/value pairs were extracted into fields, as shown in the following response:
````
"extracted into" doesn't work (we would normally use "extracted from"). Do we mean something like "were extracted and then added to fields"?

natebower replied:
Follow-up: @kolchfa-aws has indicated that "extract into" is common usage in a CS context. I would still prefer different/clearer wording but will defer to you.

vagimeli replied:
Revised to read "to find data and organize it into a structured format."

natebower reviewed (commented):
@vagimeli Please see my comments and changes and let me know if you have any questions. Thanks!

vagimeli commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+`append_separator` | Optional | The separator character or string that separates appended fields. Default is `""` (empty string).
+`description`  | Optional  | A brief description of the processor.  |
+`if` | Optional | A condition for running the processor. |
+`ignore_failure` | Optional | Specifies whether the processor continues execution even if it encounters errors. If set to `true`, failures are ignored. Default is `false`. |
```
Revised to clarify that it's the processor failure that is ignored, not the error. _If set to `true`, the processor failure is ignored._

vagimeli commented on _ingest-pipelines/processors/dissect.md:
````diff
@@ -0,0 +1,526 @@
+@timestamp: "03/Nov/2023:15:20:45 +0000"
+```
+
+The dissect pattern works by matching the string against a set of rules. For example, the first rule is to discard a single space. Dissect will find this space and then assign the value of `client_ip` to everything up to that space. The next rule is to match the `[` and `]` characters and then assign the value of `@timestamp` to everything in between.
````
Revised to read "...to all the data characters before that space."

vagimeli commented on _ingest-pipelines/processors/dissect.md:
```diff
@@ -0,0 +1,526 @@
+
+### Key modifiers 
+
+The `dissect` processor supports key modifiers that can change the default processor behavior. These modifiers are always placed to the left or right of the `%{keyname}` and are always enclosed within `%{}`. For example, the `%{+keyname->}` modifier includes the append and right padding modifiers. Key modifiers are useful for cases such as combining multiple fields into a single line of output, c ...
```
They are common names and not used in the code. The names used in the code vary, such as `append_separator` or `->`, which is the right padding modifier.

natebower replied:
Ok, can we indicate that they're names in some way? Or maybe hyphenate "right-padding"?
